### DISCUSSION TITLE
Setting up C# with omnisharp in lazyvim

### DISCUSSION THREAD

**Discussion by sample-usr** [LazyVim/LazyVim#657] [Q&A] [answered]
Hi all,
I am trying to setup `omnisharp` for csharp development. I have added it into `lua/plugins/mason.lua` like so:
```lua
return {
  {
    "williamboman/mason.nvim",
    opts = {
      ensure_installed = {
        "stylua",
        "shellcheck",
        "shfmt",
        "flake8",
        "omnisharp",
      },
    },
  },
}
```
However, omnisharp currently has an issue where I need to do a custom setup for the `on_attach` method as defined in this issue https://github.com/OmniSharp/omnisharp-roslyn/issues/2483 but I'm not sure where can I add this piece of code in lazyvim:
```lua
  if client.name == 'omnisharp' then
    local tokenModifiers = client.server_capabilities.semanticTokensProvider.legend.tokenModifiers
    for i, v in ipairs(tokenModifiers) do
      tokenModifiers[i] = v:gsub(' ', '_')
    end
    local tokenTypes = client.server_capabilities.semanticTokensProvider.legend.tokenTypes
    for i, v in ipairs(tokenTypes) do
      tokenTypes[i] = v:gsub(' ', '_')
    end
  end
```
I have tried adding it in a separate file under `plugins` like:
```lua
return {
  {
    "neovim/nvim-lspconfig",
    opts = {
      on_attach = function(client, _)
        if client.name == "omnisharp" then
          if client.name == "omnisharp" then
            local tokenModifiers = client.server_capabilities.semanticTokensProvider.legend.tokenModifiers
            for i, v in ipairs(tokenModifiers) do
              tokenModifiers[i] = v:gsub(" ", "_")
            end
            local tokenTypes = client.server_capabilities.semanticTokensProvider.legend.tokenTypes
            for i, v in ipairs(tokenTypes) do
              tokenTypes[i] = v:gsub(" ", "_")
            end
          end
        end
      end,
    },
  },
}
```
but that doesn't work.
Any suggestions?

**Comment by dpetka2001** [comment]
I don't believe `opts.on_attach` option exists. You should try `opts.setup`. There's an example in the lazyvim [recipes](https://www.lazyvim.org/configuration/recipes#fix-clangd-offset-encoding) for clangd, but it should be the same i think. Alternatively you could use an autocmd using `require("lazyvim.util").on_attach` function.

**Comment by amaanq** [comment, marked as answer]
Derive it from here:
https://github.com/amaanq/nvim-config/blob/6b05d966b2ba496b9c313221c1a80a02da8ccbfe/lua/plugins/extras/lang/csharp.lua

**Reply by sample-usr** [reply]
Thanks, that solved it. Although the shorter/general method didn't work for me and I had to revert to the long list of individual tokens (as you had it also previously). Maybe check your config again with the latest change if it's still working for you.

**Reply by amaanq** [reply]
You're right, I fixed it now - https://github.com/amaanq/nvim-config/blob/master/lua/plugins/extras/lang/csharp.lua

**Reply by MillerJ20** [reply]
@amaanq I am new to nvim, following your file format here and pasting your code. I get the error "Undefined field: server_capabilites" when I paste this code in

**Comment by sebastianstudniczek** [comment]
Isn't roslyn.nvim better nowadays than omnisharp?
